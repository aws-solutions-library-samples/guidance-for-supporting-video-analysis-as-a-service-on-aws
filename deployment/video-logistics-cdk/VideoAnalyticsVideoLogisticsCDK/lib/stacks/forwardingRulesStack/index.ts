// export * from './forwardingRulesStack';
export * from './opensearchStack';