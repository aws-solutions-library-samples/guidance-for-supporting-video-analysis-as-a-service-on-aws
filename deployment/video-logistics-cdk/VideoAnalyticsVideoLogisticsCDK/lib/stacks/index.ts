export * from './forwardingRulesStack';
export * from './schedulerStack';
export * from './timelineStack';
export * from './videoExportStack';
export * from './workflowStack';
export * from './serviceStack';
