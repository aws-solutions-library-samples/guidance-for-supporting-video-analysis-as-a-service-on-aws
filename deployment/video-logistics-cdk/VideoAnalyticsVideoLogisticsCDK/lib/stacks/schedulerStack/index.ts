export * from './openSearchPitCreation';
export * from './schedulerStack';