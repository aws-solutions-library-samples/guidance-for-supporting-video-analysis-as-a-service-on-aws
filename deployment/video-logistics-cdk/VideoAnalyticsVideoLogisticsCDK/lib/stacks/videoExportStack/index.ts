export * from './videoExportWorkflowResources';
export * from './videpExportStack';