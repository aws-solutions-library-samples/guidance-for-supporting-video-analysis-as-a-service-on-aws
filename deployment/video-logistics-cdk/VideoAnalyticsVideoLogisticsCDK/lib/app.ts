import { App } from "aws-cdk-lib";
import { 
    // ForwardingRulesStack,
    ServiceStack,
    TimelineStack,
    VideoExportStack,
    WorkflowStack
 } from "./stacks";
import { AWSRegion } from "video_analytics_common_construct";
import { VideoLogisticsBootstrapStack } from "./stacks/bootstrapStack/videoLogisticsBootstrapStack";
import { BulkInferenceStack } from "./stacks/workflowStack/bulkInferenceStack";
import { OpenSearchStack } from "./stacks/forwardingRulesStack";

const app = new App();

// Get the account and region from environment variables or use defaults
const account =
  process.env.CDK_DEPLOY_ACCOUNT ||
  process.env.CDK_DEFAULT_ACCOUNT ||
  "YOUR_DEFAULT_ACCOUNT";
const region = (process.env.CDK_DEPLOY_REGION ||
  process.env.CDK_DEFAULT_REGION ||
  "us-west-2") as AWSRegion;

// Common environment config
const envConfig = {
    env: {
        account: account,
        region: region
    },
    region: region,
    account: account
};

// Create all stacks
// new ForwardingRulesStack(app, 'VideoLogisticsForwardingRulesStack', envConfig);
// new SchedulerStack(app, 'VideoLogisticsSchedulerStack', envConfig);
// new TimelineStack(app, 'VideoLogisticsTimelineStack', envConfig);
// new VideoExportStack(app, 'VideoLogisticsVideoExportStack', envConfig);
new VideoLogisticsBootstrapStack(app, 'VideoLogisticsBootstrapStack', envConfig);
let opensearchStack = new OpenSearchStack(app, 'VideoLogisticsOpensearchStack', envConfig);
new BulkInferenceStack(app, 'VideoLogisticsBulkInferenceStack', {...envConfig, opensearchEndpoint: opensearchStack.opensearchEndpoint});
new WorkflowStack(app, 'VideoLogisticsWorkflowStack', envConfig);
new ServiceStack(app, 'VideoLogisticsServiceStack', envConfig);

app.synth();
