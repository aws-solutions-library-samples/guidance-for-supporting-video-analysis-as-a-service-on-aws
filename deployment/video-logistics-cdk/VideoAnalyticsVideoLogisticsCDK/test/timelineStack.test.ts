import { createStackSnapshotTest } from "./snapshotTestUtil";
import { TimelineStack } from "../lib/stacks";

createStackSnapshotTest("TimelineStack", TimelineStack);