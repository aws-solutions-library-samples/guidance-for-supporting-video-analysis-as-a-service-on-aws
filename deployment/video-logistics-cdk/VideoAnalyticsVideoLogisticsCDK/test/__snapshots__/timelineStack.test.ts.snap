// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineStack Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "RawVideoTimelineTableA7CCDA88": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "CustomerId_DeviceId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "Timestamp",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "CustomerId_DeviceId",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "Timestamp",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableClass": "STANDARD",
        "TableName": "RawVideoTimelineTable",
        "TimeToLiveSpecification": {
          "AttributeName": "ExpirationTimestamp",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "RawVideoTimelineTableStreamDLQ506E0C04": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "TimelineStreamDlqEncryptionKey654F7F66",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "S3ExporterLambdaRole1D87DB8D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Allows lambda to export timeline objects in S3 to DDB",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "S3ExporterLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "S3ExporterLambdaRoleDefaultPolicyBB437BDC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:DeleteObject",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::videoanalytics-timeline-bucket-us-east-1-123456789012",
                "arn:aws:s3:::videoanalytics-timeline-bucket-us-east-1-123456789012/*",
              ],
            },
            {
              "Action": [
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RawVideoTimelineTableA7CCDA88",
                  "Arn",
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineBucketEncryptionKeyFE3340B2",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3ExporterLambdaRoleDefaultPolicyBB437BDC",
        "Roles": [
          {
            "Ref": "S3ExporterLambdaRole1D87DB8D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "S3ExporterXMinutesRule3B8D624A": {
      "Properties": {
        "ScheduleExpression": "cron(0/1 * * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "VideoTimelineS3ExportLambda5BBEB50A",
                "Arn",
              ],
            },
            "Id": "Target0",
            "RetryPolicy": {
              "MaximumEventAgeInSeconds": 60,
              "MaximumRetryAttempts": 1,
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "S3ExporterXMinutesRuleAllowEventRuleTimelineStackVideoTimelineS3ExportLambdaC56263CC0F3C69BC": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "VideoTimelineS3ExportLambda5BBEB50A",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "S3ExporterXMinutesRule3B8D624A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TimelineBucketA0B105A7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "KMSMasterKeyID": {
                  "Fn::GetAtt": [
                    "TimelineBucketEncryptionKeyFE3340B2",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "videoanalytics-timeline-bucket-us-east-1-123456789012",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 90,
              "Id": "ExpireAfterNinetyDays",
              "NoncurrentVersionExpiration": {
                "NoncurrentDays": 90,
              },
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogs56A4C014",
          },
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TimelineBucketEncryptionKeyFE3340B2": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TimelineBucketPolicyAC1D19C6": {
      "Properties": {
        "Bucket": {
          "Ref": "TimelineBucketA0B105A7",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TimelineBucketA0B105A7",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TimelineBucketA0B105A7",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TimelineForwarderLambda2F7F49AF": {
      "DependsOn": [
        "TimelineForwarderLambdaRoleDefaultPolicyA01D88FB",
        "TimelineForwarderLambdaRoleEF547331",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "49a15331380e380f203696c106711a7a83279bf39949782b80dd733783d209f3.jar",
        },
        "Description": "Lambda responsible for forwarding lambda to KDS ",
        "Environment": {
          "Variables": {
            "rawPartitionKey": "CustomerId_DeviceId",
            "rawSortKey": "Timestamp",
            "table": "RawVideoTimelineTable",
          },
        },
        "Handler": "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.TimelineForwarderLambda::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "TimelineForwarderLambdaLogGroupBA513DCE",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "TimelineForwarderLambdaRoleEF547331",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 720,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TimelineForwarderLambdaDLQ932671BA": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "TimelineKey51519BA3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "TimelineForwarderLambdaDynamoDBEventSourceTimelineStackRawVideoTimelineTable47709EDEE6EBD00B": {
      "Properties": {
        "BatchSize": 10,
        "BisectBatchOnFunctionError": true,
        "DestinationConfig": {
          "OnFailure": {
            "Destination": {
              "Fn::GetAtt": [
                "TimelineForwarderLambdaDLQ932671BA",
                "Arn",
              ],
            },
          },
        },
        "Enabled": true,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "RawVideoTimelineTableA7CCDA88",
            "StreamArn",
          ],
        },
        "FunctionName": {
          "Ref": "TimelineForwarderLambda2F7F49AF",
        },
        "MaximumRetryAttempts": 2,
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TimelineForwarderLambdaLogGroupBA513DCE": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "TimelineForwarderLambdaLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TimelineForwarderLambdaRoleDefaultPolicyA01D88FB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineForwarderLambdaDLQ932671BA",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineKey51519BA3",
                  "Arn",
                ],
              },
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RawVideoTimelineTableA7CCDA88",
                  "StreamArn",
                ],
              },
            },
            {
              "Action": "kinesis:PutRecord",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:GetRecords",
                "dynamodb:GetItem",
                "dynamodb:Query",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": [
                "*",
                {
                  "Fn::GetAtt": [
                    "RawVideoTimelineTableA7CCDA88",
                    "Arn",
                  ],
                },
                {
                  "Fn::GetAtt": [
                    "VideoTimelineTableBE74FCCA",
                    "Arn",
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TimelineForwarderLambdaRoleDefaultPolicyA01D88FB",
        "Roles": [
          {
            "Ref": "TimelineForwarderLambdaRoleEF547331",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TimelineForwarderLambdaRoleEF547331": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Allows lambda to forward DDB requests to KDS",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "TmelineForwarderLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "TimelineKDSB3E27D47": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Name": "TimelineKDS",
        "RetentionPeriodHours": 168,
        "StreamEncryption": {
          "EncryptionType": "KMS",
          "KeyId": {
            "Fn::GetAtt": [
              "TimelineKey51519BA3",
              "Arn",
            ],
          },
        },
        "StreamModeDetails": {
          "StreamMode": "ON_DEMAND",
        },
      },
      "Type": "AWS::Kinesis::Stream",
      "UpdateReplacePolicy": "Retain",
    },
    "TimelineKey51519BA3": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TimelineStreamDlqEncryptionKey654F7F66": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "VideoDensityUpdateLambda4FEC9CBA": {
      "DependsOn": [
        "VideoTimelineLambdaRoleDefaultPolicy976A2586",
        "VideoTimelineLambdaRoleB013BF74",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "49a15331380e380f203696c106711a7a83279bf39949782b80dd733783d209f3.jar",
        },
        "Description": "Lambda responsible for aggregation of video timelines",
        "Environment": {
          "Variables": {
            "rawPartitionKey": "CustomerId_DeviceId",
            "rawSortKey": "Timestamp",
            "table": "RawVideoTimelineTable",
          },
        },
        "Handler": "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoDensityUpdateLambda::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "VideoTimelineLambdaLogGroup59268093",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "VideoTimelineLambdaRoleB013BF74",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 720,
      },
      "Type": "AWS::Lambda::Function",
    },
    "VideoDensityUpdateLambdaKinesisEventSourceTimelineStackTimelineKDSBE52AAAEF0CE43C8": {
      "Properties": {
        "BatchSize": 200,
        "BisectBatchOnFunctionError": false,
        "DestinationConfig": {
          "OnFailure": {
            "Destination": {
              "Fn::GetAtt": [
                "RawVideoTimelineTableStreamDLQ506E0C04",
                "Arn",
              ],
            },
          },
        },
        "Enabled": true,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "TimelineKDSB3E27D47",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "VideoDensityUpdateLambda4FEC9CBA",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
        "MaximumBatchingWindowInSeconds": 3,
        "MaximumRecordAgeInSeconds": 604800,
        "MaximumRetryAttempts": 1,
        "ParallelizationFactor": 5,
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "VideoTimelineLambdaLogGroup59268093": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "VideoTimelineLambdaLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "VideoTimelineLambdaRoleB013BF74": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "VideoTimelineLambdaRoleDefaultPolicy976A2586": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineKey51519BA3",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "dynamodb:GetRecords",
                "dynamodb:GetItem",
                "dynamodb:Query",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": [
                "*",
                {
                  "Fn::GetAtt": [
                    "RawVideoTimelineTableA7CCDA88",
                    "Arn",
                  ],
                },
                {
                  "Fn::GetAtt": [
                    "VideoTimelineTableBE74FCCA",
                    "Arn",
                  ],
                },
              ],
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RawVideoTimelineTableStreamDLQ506E0C04",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineStreamDlqEncryptionKey654F7F66",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
                "kinesis:DescribeStream",
                "kinesis:ListStreams",
                "kinesis:DescribeStreamConsumer",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineKDSB3E27D47",
                  "Arn",
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineKey51519BA3",
                  "Arn",
                ],
              },
            },
            {
              "Action": "kinesis:DescribeStream",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TimelineKDSB3E27D47",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VideoTimelineLambdaRoleDefaultPolicy976A2586",
        "Roles": [
          {
            "Ref": "VideoTimelineLambdaRoleB013BF74",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VideoTimelineS3ExportLambda5BBEB50A": {
      "DependsOn": [
        "S3ExporterLambdaRoleDefaultPolicyBB437BDC",
        "S3ExporterLambdaRole1D87DB8D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "49a15331380e380f203696c106711a7a83279bf39949782b80dd733783d209f3.jar",
        },
        "Description": "Lambda responsible for exporting timeline information from S3 to DDB",
        "Environment": {
          "Variables": {
            "AWS_ACCOUNT": "123456789012",
          },
        },
        "Handler": "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoTimelineS3ExportLambda::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "VideoTimelineS3ExportLambdaLogGroup72F201EA",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "S3ExporterLambdaRole1D87DB8D",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 720,
      },
      "Type": "AWS::Lambda::Function",
    },
    "VideoTimelineS3ExportLambdaLogGroup72F201EA": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "VideoTimelineS3ExportLambdaLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "VideoTimelineTableBE74FCCA": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "CustomerId_DeviceId_TimeUnit",
            "AttributeType": "S",
          },
          {
            "AttributeName": "UnitTimestamp",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "CustomerId_DeviceId_TimeUnit",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "UnitTimestamp",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableClass": "STANDARD",
        "TableName": "VideoTimelineTable",
        "TimeToLiveSpecification": {
          "AttributeName": "ExpirationTimestamp",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogs56A4C014": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogsPolicy394FF133": {
      "Properties": {
        "Bucket": {
          "Ref": "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogs56A4C014",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogs56A4C014",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "videoanalyticstimelinebucketuseast1123456789012ServerAccessLogs56A4C014",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
